### ISSUE TITLE
Persistent extremely loud clicking noise when opening on linux

### ISSUE THREAD

**Issue by randomouscrap98** [JohnEarnest/Decker#78]
I built the program on arch linux using the latest sdl2 and sdl2_image. Everything built fine; no warnings or anything. The program itself runs great, there doesn't seem to be any issues. However, there is a persistent, extremely loud clicking playing constantly, no matter what file I load. It starts as soon as the program starts, and whether I do a new deck or load a file or anything, the sound plays continuously. Other sounds within the program work, they just play alongside the presumably) max volume clicking sound.

I opened the waveform in audacity and this is what it looks like:
![Screenshot_2024-02-20_02-22-10](https://github.com/JohnEarnest/Decker/assets/13357277/487322c4-bb2e-4b02-8ef3-85dcef021f8a)

Here's the sound file if you're interested:
[out.zip](https://github.com/JohnEarnest/Decker/files/14340878/out.zip)

If you need more information, please let me know. I might try to reproduce the issue on other hardware and distros if possible. I used the latest build as of the time of writing (I just cloned it an hour or two ago).

**Comment by JohnEarnest** [comment]
Off the top of my head this sounds like signed/unsigned conversion shenanigans somewhere in the audio pump highlighting differences in behavior between clang and gcc.

This looks rather suspicious to me:
https://github.com/JohnEarnest/Decker/blob/2fff47e4e70782685de6acc530d4f9ffcebecda8/c/decker.c#L3047-L3048

I _think_ it would be more correct to do something like
```c
int b=(signed char)audio_slots[z].clip->sv[audio_slots[z].sample];
```

**Comment by randomouscrap98** [comment]
I'm sorry, that didn't seem to work. I made the change and very carefully checked it 3 times, and cleaned/built it twice to make sure. I can try other things if you'd like. Just to be clear: I replaced the two lines given with the one line as shown.

**Comment by randomouscrap98** [comment]
I went ahead and looked a little closer at the waveform and found that it's not quite one blip, it's spread out. I don't know if this is FFT stuff or whatever, or some kind of post processing. I'm not super familiar with audio.

![Screenshot_2024-02-20_03-53-03](https://github.com/JohnEarnest/Decker/assets/13357277/386f8115-ee04-4e15-97b1-245985a6ed91)

**Comment by randomouscrap98** [comment]
I compiled with clang rather than gcc and the sound clicking persisted. I tried both versions; the original and the suggested change. The versions of the SDL libraries I'm using are:
`SDL2: 2.30.0-1` (released 3 weeks ago, see https://github.com/libsdl-org/SDL/releases/tag/release-2.30.0)
`SDL2_Image: 2.8.2-2` (released January 1, see https://github.com/libsdl-org/SDL_image/releases/tag/release-2.8.2)

Sorry for all the messages, just hoping it helps

**Comment by randomouscrap98** [comment]
I removed the body of `n_play` and `sfx_install` and the clicking persists. I verified that the sound system was basically non-functional by opening decks with sounds (or just trying to play sounds from the menu). Strangely though, removing the body of `sfx_pump` also appears to do nothing. I'm wondering if perhaps this is just a glitch with SDL with my particular hardware. I'll try building this on some other hardware when I can. It should be noted that running with `--no-sound` does indeed remove the clicking.

**Comment by randomouscrap98** [comment]
Hm I think you're right about a regression.

I tried it on another arch system, completely different hardware, and it still made the clicking. I tried swapping the audio hardware to use an external device and it still did it. But when I tried on ubuntu using sdl 2.0.20 (the version available on the system I'm using) there were no clicks.

I tested out a very basic example program that only creates an audio context and should (theoretically) play silence for 5 seconds. It also clicks on the arch systems, so this is either a bug in SDL or perhaps an invalid configuration. I noticed that the clicks in my example program don't happen unless I set the format to `AUDIO_S8`. I'll go look around at the documentation to see if there's anything mentioning something about this, then I'll make an issue on SDL if not. 

Thank you for your suggestions! I understand if you want to close the issue, but I'd really like to wait until I've looked over the documentation on the slim chance that this isn't a bug on their end. Up to you of course.

**Comment by JohnEarnest** [comment]
Let's definitely keep this issue open until we're certain it's *entirely* an SDL problem.

**Comment by randomouscrap98** [comment]
The code mentioned previously is:

```
#include <SDL.h>
#include <SDL_image.h>

void audio_callback(void *userdata, Uint8 *stream, int len) {
    // Fill the audio stream with zeros
    memset(stream, 0, len);
}

int main() {
    // Initialize SDL
    if (SDL_Init(SDL_INIT_AUDIO) < 0) {
        // Handle initialization error
        return -1;
    }

    // Set the audio format
    SDL_AudioSpec audio_spec;
    audio_spec.freq = 8000;
    audio_spec.format = AUDIO_S8; //AUDIO_S16SYS;
    audio_spec.channels = 2;
    audio_spec.samples = 1024;
    audio_spec.callback = audio_callback;

    // Open the audio device
    if (SDL_OpenAudio(&audio_spec, NULL) < 0) {
        // Handle audio device opening error
        return -1;
    }

    // Start playing audio
    SDL_PauseAudio(0);

    // Wait for some time (or perform other tasks)
    SDL_Delay(5000); // Wait for 5 seconds

    // Pause audio playback
    SDL_PauseAudio(1);

    // Keep the audio device open but stop playback
    // The audio_callback function will continue to fill the stream with zeros

    // Wait for user input or perform other cleanup tasks
    // ...

    // Close the audio device and quit SDL
    SDL_CloseAudio();
    SDL_Quit();

    return 0;
}
```

Built with `gcc $(sdl2-config --cflags --libs) -lSDL2_image main.c` (it doesn't matter if I use gcc or clang). I included the image library just to make it as similar as possible, though I assume none of it is used. I tried it without the image library and it's still the same.

**Comment by randomouscrap98** [comment]
I've done some testing and research on various examples and past issues, including looking for examples that are supposed to work with specifically 2.30. According to the examples, they didn't change anything, and I'm not seeing anything we could be doing wrong, either with Decker or the super small example program. I've created an issue on SDL, though it might be an arch-specific issue. I built SDL 2.30 from source and tried that and it still produces the clicking on the simple program. Hopefully we'll see what happens, but I don't think Decker is doing anything wrong: https://github.com/libsdl-org/SDL/issues/9099
